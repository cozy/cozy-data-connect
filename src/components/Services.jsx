import React from 'react'
import PropTypes from 'prop-types'
import sortBy from 'lodash/sortBy'
import { connect } from 'react-redux'
import { withClient } from 'cozy-client'
import { queryConnect } from 'cozy-client'
import { translate } from 'cozy-ui/react/I18n'
import keyBy from 'lodash/keyBy'
import has from 'lodash/has'
import flow from 'lodash/flow'

import AddServiceTile from 'components/AddServiceTile'
import KonnectorTile from 'components/KonnectorTile'
import CandidateCategoryTile from 'components/CandidateCategoryTile'
import CandidateServiceTile from 'components/CandidateServiceTile'
import EmptyServicesListTip from 'components/EmptyServicesListTip'
import { getInstalledKonnectors } from 'reducers/index'
import useAppsInMaintenance from 'hooks/withAppsInMaintenance'
import candidatesConfig from 'config/candidates'

export const Services = ({
  t,
  installedKonnectors,
  suggestedKonnectorsQuery,
  client
}) => {
  const hasConnections = !!installedKonnectors.length
  const appsInMaintenance = useAppsInMaintenance(client)
  const appsInMaintenanceBySlug = keyBy(appsInMaintenance, 'slug')
  const suggestedKonnectors = suggestedKonnectorsQuery.data

  return (
    <>
      <div className={'services-list'}>
        {installedKonnectors.map((konnector, index) => (
          <KonnectorTile
            key={index}
            konnector={konnector}
            route={`connected/${konnector.slug}`}
            isInMaintenance={has(appsInMaintenanceBySlug, konnector.slug)}
          />
        ))}
        {!hasConnections &&
          suggestedKonnectors.length === 0 &&
          candidatesConfig.konnectors.map(candidate => (
            <CandidateServiceTile key={candidate.slug} konnector={candidate} />
          ))}
        {suggestedKonnectors
          // TODO turn this into a method on the model
          .filter(
            ({ slug }) =>
              !has(appsInMaintenanceBySlug, slug) &&
              !installedKonnectors.some(
                installedKonnector => installedKonnector.slug === slug
              )
          )
          .map(suggestion => (
            <CandidateServiceTile
              key={suggestion.slug}
              konnector={suggestion}
            />
          ))}
        {!hasConnections &&
          Object.entries(candidatesConfig.categories).map(
            ([category, slugs]) => (
              <CandidateCategoryTile
                key={category}
                slugs={slugs}
                category={category}
              />
            )
          )}
        {<AddServiceTile label={t('add_service')} />}
      </div>
      {!hasConnections &&
        suggestedKonnectors.length >= 1 && <EmptyServicesListTip />}
    </>
  )
}

Services.propTypes = {
  t: PropTypes.func.isRequired,
  installedKonnectors: PropTypes.arrayOf(
    PropTypes.shape({ slug: PropTypes.string })
  ).isRequired,
  suggestedKonnectorsQuery: PropTypes.shape({
    data: PropTypes.array.isRequired
  }).isRequired,
  client: PropTypes.object.isRequired
}

const query = client =>
  client
    .find('io.cozy.apps.suggestions')
    .where({ silenced: false })
    .sortBy([{ silenced: 'asc' }, { slug: 'asc' }])
    .indexFields(['silenced', 'slug'])

const mapStateToProps = state => {
  return {
    installedKonnectors: sortBy(getInstalledKonnectors(state), konnector =>
      konnector.name.toLowerCase()
    )
  }
}

export default flow(
  connect(mapStateToProps),
  translate(),
  queryConnect({ suggestedKonnectorsQuery: { query } }),
  withClient
)(Services)
