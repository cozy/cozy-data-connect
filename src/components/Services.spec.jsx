'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'
import { Services } from './Services'
import { tMock } from '../../test/jestLib/I18n'

jest.mock('cozy-ui/transpiled/react/utils/color', () => ({
  getCssVariableValue: () => '#fff'
}))

describe('Services component', () => {
  it('should display a list of services', () => {
    const installedKonnectors = [
      { slug: 'test1' },
      { slug: 'test2' },
      { slug: 'test3' }
    ]
    const component = shallow(
      <Services
        t={tMock}
        installedKonnectors={installedKonnectors}
        suggestedKonnectorsQuery={{ data: [] }}
        client={{}}
      />
    )
    expect(component.getElement()).toMatchSnapshot()
  })

  it('should display the empty component when there are no services', () => {
    const installedKonnectors = []
    const component = shallow(
      <Services
        t={tMock}
        installedKonnectors={installedKonnectors}
        suggestedKonnectorsQuery={{ data: [{ slug: 'suggestion-1' }] }}
        client={{}}
      />
    )
    expect(component.getElement()).toMatchSnapshot()
    const emptyServicesListTip = component
      .children()
      .last()
      .dive()
    expect(emptyServicesListTip.getElement()).toMatchSnapshot()
  })

  it('should show default suggestions when there are no services and no suggestions', () => {
    const component = shallow(
      <Services
        t={tMock}
        installedKonnectors={[]}
        suggestedKonnectorsQuery={{ data: [] }}
        client={{}}
      />
    )
    expect(component.getElement()).toMatchSnapshot()
  })

  it('should display suggestions after installed services', () => {
    const installedKonnectors = [
      { slug: 'test-1' },
      { slug: 'test-2' },
      { slug: 'test-3' }
    ]
    const component = shallow(
      <Services
        t={tMock}
        installedKonnectors={installedKonnectors}
        suggestedKonnectorsQuery={{
          data: [
            { slug: 'suggestion-1' },
            { slug: 'suggestion-2' },
            { slug: 'test-1' }
          ]
        }}
        client={{}}
      />
    )
    expect(component.getElement()).toMatchSnapshot()
  })
})
