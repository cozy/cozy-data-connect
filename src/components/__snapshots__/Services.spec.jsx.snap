// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Services component should display a list of services 1`] = `
<React.Fragment>
  <div
    className="services-list"
  >
    <Wrapper
      isInMaintenance={false}
      konnector={
        Object {
          "slug": "test1",
        }
      }
      route="connected/test1"
    />
    <Wrapper
      isInMaintenance={false}
      konnector={
        Object {
          "slug": "test2",
        }
      }
      route="connected/test2"
    />
    <Wrapper
      isInMaintenance={false}
      konnector={
        Object {
          "slug": "test3",
        }
      }
      route="connected/test3"
    />
    <withClient(AddServiceTile)
      label="Add"
    />
  </div>
</React.Fragment>
`;

exports[`Services component should display suggestions after installed services 1`] = `
<React.Fragment>
  <div
    className="services-list"
  >
    <Wrapper
      isInMaintenance={false}
      konnector={
        Object {
          "slug": "test-1",
        }
      }
      route="connected/test-1"
    />
    <Wrapper
      isInMaintenance={false}
      konnector={
        Object {
          "slug": "test-2",
        }
      }
      route="connected/test-2"
    />
    <Wrapper
      isInMaintenance={false}
      konnector={
        Object {
          "slug": "test-3",
        }
      }
      route="connected/test-3"
    />
    <Wrapper
      konnector={
        Object {
          "slug": "suggestion-1",
        }
      }
    />
    <Wrapper
      konnector={
        Object {
          "slug": "suggestion-2",
        }
      }
    />
    <withClient(AddServiceTile)
      label="Add"
    />
  </div>
</React.Fragment>
`;

exports[`Services component should display the empty component when there are no services 1`] = `
<React.Fragment>
  <div
    className="services-list"
  >
    <Wrapper
      konnector={
        Object {
          "slug": "suggestion-1",
        }
      }
    />
    <Wrapper
      category="banking"
      slugs={
        Array [
          "hsbc119",
          "caissedepargne1",
          "ingdirect95",
          "cic45",
        ]
      }
    />
    <Wrapper
      category="isp"
      slugs={
        Array [
          "sfr",
          "free",
          "orange",
          "bouyguestelecom",
        ]
      }
    />
    <Wrapper
      category="insurance"
      slugs={
        Array [
          "maifecheancier",
          "alan",
          "harmonie",
          "macif",
        ]
      }
    />
    <Wrapper
      category="energy"
      slugs={
        Array [
          "engie",
          "planeteoui",
          "veoliaeau",
          "ekwateur",
        ]
      }
    />
    <withClient(AddServiceTile)
      label="Add"
    />
  </div>
  <Wrapper />
</React.Fragment>
`;

exports[`Services component should display the empty component when there are no services 2`] = `<EmptyServicesListTip />`;

exports[`Services component should show default suggestions when there are no services and no suggestions 1`] = `
<React.Fragment>
  <div
    className="services-list"
  >
    <Wrapper
      slug="ameli"
    />
    <Wrapper
      slug="impots"
    />
    <Wrapper
      category="banking"
      slugs={
        Array [
          "hsbc119",
          "caissedepargne1",
          "ingdirect95",
          "cic45",
        ]
      }
    />
    <Wrapper
      category="isp"
      slugs={
        Array [
          "sfr",
          "free",
          "orange",
          "bouyguestelecom",
        ]
      }
    />
    <Wrapper
      category="insurance"
      slugs={
        Array [
          "maifecheancier",
          "alan",
          "harmonie",
          "macif",
        ]
      }
    />
    <Wrapper
      category="energy"
      slugs={
        Array [
          "engie",
          "planeteoui",
          "veoliaeau",
          "ekwateur",
        ]
      }
    />
    <withClient(AddServiceTile)
      label="Add"
    />
  </div>
  <Wrapper />
</React.Fragment>
`;
